### PR TITLE
[NOD-1414] Write domain interfaces and stub implementations for the new kaspadstate architecture

### PR DESCRIPTION
**Pull request by stasatdaglabs**

### CODE REVIEW

File: domain/kaspadstate/algorithms/interfaces.go [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+	ValidateBodyInIsolation(block *appmessage.MsgBlock) error
+	ValidateBodyInContext(block *appmessage.MsgBlock) error
+}
+
+// ConsensusStateManager ...
+type ConsensusStateManager interface {
+	UTXOByOutpoint(outpoint *appmessage.Outpoint) *model.UTXOEntry
+	ValidateTransaction(transaction *util.Tx, utxoEntries []*model.UTXOEntry) error
+
+	SerializedUTXOSet() []byte
```
Comment (svarogg) [inline domain/kaspadstate/algorithms/interfaces.go:30]: Should move to pruning

File: domain/kaspadstate/algorithms/interfaces.go [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+	CalculateConsensusStateChanges(block *appmessage.MsgBlock) *model.ConsensusStateChanges
+}
+
+// DAGTopologyManager ...
+type DAGTopologyManager interface {
+	Parents(blockHash *daghash.Hash) []*daghash.Hash
+	Children(blockHash *daghash.Hash) []*daghash.Hash
```
Comment (svarogg) [inline domain/kaspadstate/algorithms/interfaces.go:37]: Children misses from UML. Other methods  have different signatures in UML

File: domain/kaspadstate/algorithms/interfaces.go [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+
+// GHOSTDAGManager ...
+type GHOSTDAGManager interface {
+	GHOSTDAG(blockParents []*daghash.Hash) *model.BlockGHOSTDAGData
+	BlockData(blockHash *daghash.Hash) *model.BlockGHOSTDAGData
+}
+
+// PruningManager ...
+type PruningManager interface {
+	FindPruningPoint(blockHash *daghash.Hash) *daghash.Hash
```
Comment (svarogg) [inline domain/kaspadstate/algorithms/interfaces.go:58]: Should be modified as discusses

File: domain/kaspadstate/algorithms/interfaces.go [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+}
+
+// ReachabilityTree ...
+type ReachabilityTree interface {
+	AddNode(dbTx *dbaccess.TxContext, blockHash *daghash.Hash)
+	IsInPastOf(blockHashA *daghash.Hash, blockHashB *daghash.Hash) bool
```
Comment (svarogg) [inline domain/kaspadstate/algorithms/interfaces.go:64]: Rename: IsInPast ->  IsReachableFrom
Add: IsReachabilityTreeAncestorOf
Make DAGTopology use those for isInPast/isInSelectedParentChain. Update dependancies accordingly

File: domain/kaspadstate/model/readonlyutxoset.go [hunk 5]
```diff
@@ -0,0 +1,5 @@
+package model
+
+// ReadOnlyUTXOSet ...
+type ReadOnlyUTXOSet interface {
```
Comment (svarogg) [inline domain/kaspadstate/model/readonlyutxoset.go:4]: We should decide what's the API of ReadOnlyUTXOSet, SelectedParentIterator, etc.

Someone is going to implement those. Someone else is going to use those.
Therefore APIs should be ready ahead of time.

**Review by svarogg — CHANGES_REQUESTED** [review]

**Review by someone235 — CHANGES_REQUESTED** [review]

File: domain/kaspadstate/algorithms/ghostdagmanager/ghostdagmanager.go [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+	dagTopologyManager algorithms.DAGTopologyManager,
+	ghostdagDataStore datastructures.GHOSTDAGDataStore) *GHOSTDAGManager {
+	return &GHOSTDAGManager{
+		dagTopologyManager: dagTopologyManager,
+		ghostdagDataStore:  ghostdagDataStore,
+	}
+}
+
+// GHOSTDAG ...
+func (gm *GHOSTDAGManager) GHOSTDAG(blockParents []*daghash.Hash) *model.BlockGHOSTDAGData {
```
[the remark this answers is not in the document]
Comment (stasatdaglabs) [reply domain/kaspadstate/algorithms/ghostdagmanager/ghostdagmanager.go:27]: I don't want to save its data, though.

File: domain/kaspadstate/algorithms/interfaces.go [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+	"github.com/kaspanet/kaspad/app/appmessage"
+	"github.com/kaspanet/kaspad/domain/kaspadstate/model"
+	"github.com/kaspanet/kaspad/infrastructure/db/dbaccess"
+	"github.com/kaspanet/kaspad/util"
+	"github.com/kaspanet/kaspad/util/daghash"
+)
+
+// BlockProcessor ...
+type BlockProcessor interface {
+	BuildBlock(transactionSelector model.TransactionSelector) *appmessage.MsgBlock
```
[the remark this answers is not in the document]
Comment (stasatdaglabs) [reply domain/kaspadstate/algorithms/interfaces.go:13]: This is the external API. You're thinking of the internal buildBlock.

File: domain/kaspadstate/datastructures/blockstatusstore/blockstatusstore.go [hunk 8]
```diff
@@ -0,0 +1,26 @@
+package blockstatusstore
+
+import (
+	"github.com/kaspanet/kaspad/domain/kaspadstate/model"
+	"github.com/kaspanet/kaspad/infrastructure/db/dbaccess"
+	"github.com/kaspanet/kaspad/util/daghash"
+)
+
+// BlockStatusStore ...
```
[the remark this answers is not in the document]
Comment (someone235) [reply domain/kaspadstate/datastructures/blockstatusstore/blockstatusstore.go:9]: Real comments are still missing

File: domain/kaspadstate/algorithms/interfaces.go [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+	ValidateAndInsertBlock(block *appmessage.MsgBlock) error
+}
+
+// BlockValidator ...
+type BlockValidator interface {
+	ValidateHeaderInIsolation(block *appmessage.MsgBlock) error
+	ValidateHeaderInContext(block *appmessage.MsgBlock) error
+	ValidateBodyInIsolation(block *appmessage.MsgBlock) error
+	ValidateBodyInContext(block *appmessage.MsgBlock) error
```
Comment (svarogg) [inline domain/kaspadstate/algorithms/interfaces.go:21]: You are missing ValidateWithUTXO

File: domain/miningmanager/blocktemplatebuilder/blocktemplatebuilder.go [hunk 10]
```diff
@@ -0,0 +1,23 @@
...
+
+// New creates a new BlockTemplateBuilder
+func New(kaspadState *kaspadstate.KaspadState) *BlockTemplateBuilder {
+	return &BlockTemplateBuilder{
+		kaspadState: kaspadState,
+	}
+}
+
+// GetBlockTemplate creates a block template for a miner to consume
+func (btb *BlockTemplateBuilder) GetBlockTemplate() *appmessage.MsgBlock {
```
Comment (svarogg) [inline domain/miningmanager/blocktemplatebuilder/blocktemplatebuilder.go:21]: Should receive `scriptPubKey []byte, extraData []byte` as discussed

**Review by svarogg — COMMENTED** [review]
1. Didn't we say we rename "algorithms" to "processes"?
2. Please complete all the comments. Comment is like spec for the implementer. Otherwise we are going to get misunderstandings regarding what people should implement.
3. Please fill-in the fields of the models. Otherwise it will be impossible to implement some of the processes"

File: domain/consensus/processes/interfaces.go [hunk 11]
```diff
@@ -0,0 +1,73 @@
...
+	ValidateAgainstPastUTXO(block *appmessage.MsgBlock) error
+	ValidateFinality(block *appmessage.MsgBlock) error
+}
+
+// ConsensusStateManager manages the node's consensus state
+type ConsensusStateManager interface {
+	UTXOByOutpoint(outpoint *appmessage.Outpoint) *model.UTXOEntry
+	ValidateTransaction(transaction *util.Tx, utxoEntries []*model.UTXOEntry) error
+
+	CalculateConsensusStateChanges(block *appmessage.MsgBlock) *model.ConsensusStateChanges
```
Comment (svarogg) [inline domain/consensus/processes/interfaces.go:33]: Please add ResolveFinalityConflict(finalityBlockHash) both here and ExternalAPI

**Review by svarogg — CHANGES_REQUESTED** [review]
1. How do we handle notifications?
2. (I think this one we want to defer to closer to Pruning implementation, but writing down just-in-case): For Pruning we'll need some xxxDataStore.DeleteBlabla() methods.

File: domain/consensus/processes/interfaces.go [hunk 12]
```diff
@@ -0,0 +1,73 @@
...
+	IsParentOf(blockHashA *daghash.Hash, blockHashB *daghash.Hash) bool
+	IsChildOf(blockHashA *daghash.Hash, blockHashB *daghash.Hash) bool
+	IsAncestorOf(blockHashA *daghash.Hash, blockHashB *daghash.Hash) bool
+	IsDescendantOf(blockHashA *daghash.Hash, blockHashB *daghash.Hash) bool
+}
+
+// DAGTraversalManager exposes methods for travering blocks
+// in the DAG
+type DAGTraversalManager interface {
+	BlockAtDepth(uint64) *daghash.Hash
```
Comment (svarogg) [inline domain/consensus/processes/interfaces.go:50]: Should also receive block hash

**Review by svarogg — CHANGES_REQUESTED** [review]

**Review by svarogg — CHANGES_REQUESTED** [review]

**Review by svarogg — APPROVED** [review]

File: domain/consensus/model/reachabilitychangeset.go [hunk 13]
```diff
@@ -0,0 +1,10 @@
+package model
+
+import "github.com/kaspanet/kaspad/util/daghash"
+
+// ReachabilityChangeset holds the set of changes to make to a
+// reachability tree to insert a new reachability node
+type ReachabilityChangeset struct {
+	treeNodeChanges          map[*daghash.Hash]*reachabilityTreeNode
```
Comment (someone235) [inline domain/consensus/model/reachabilitychangeset.go:8]: all fields should be public, no?

File: domain/consensus/model/reachabilitydata.go [hunk 14]
```diff
@@ -0,0 +1,61 @@
+package model
+
+// ReachabilityData holds the set of data required to answer
+// reachability queries
+type ReachabilityData struct {
+	treeNode          *reachabilityTreeNode
```
Comment (someone235) [inline domain/consensus/model/reachabilitydata.go:6]: all fields should be public, no?

File: domain/consensus/model/reachabilitydata.go [hunk 15]
```diff
@@ -0,0 +1,61 @@
...
+// The main challenge of maintaining such intervals is that our tree
+// is an ever-growing tree and as such pre-allocated intervals may
+// not suffice as per future events. This is where the reindexing
+// algorithm below comes into place.
+// We use the reasonable assumption that the initial root interval
+// (e.g., [0, 2^64-1]) should always suffice for any practical use-
+// case, and so reindexing should always succeed unless more than
+// 2^64 blocks are added to the DAG/tree.
+type reachabilityTreeNode struct {
+	children []*reachabilityTreeNode
```
Comment (someone235) [inline domain/consensus/model/reachabilitydata.go:26]: all fields should be public, no?

File: domain/consensus/model/reachabilitydata.go [hunk 16]
```diff
@@ -0,0 +1,61 @@
...
+	// interval is the index interval containing all intervals of
+	// blocks in this node's subtree
+	interval *reachabilityInterval
+}
+
+// reachabilityInterval represents an interval to be used within the
+// tree reachability algorithm. See reachabilityTreeNode for further
+// details.
+type reachabilityInterval struct {
+	start uint64
```
Comment (someone235) [inline domain/consensus/model/reachabilitydata.go:38]: all fields should be public, no?

File: domain/consensus/model/reachabilitydata.go [hunk 17]
```diff
@@ -0,0 +1,61 @@
...
+	end   uint64
+}
+
+// orderedTreeNodeSet is an ordered set of reachabilityTreeNodes
+// Note that this type does not validate order validity. It's the
+// responsibility of the caller to construct instances of this
+// type properly.
+type orderedTreeNodeSet []*reachabilityTreeNode
```
Comment (someone235) [inline domain/consensus/model/reachabilitydata.go:46]: the type should be public, no?

File: domain/consensus/model/reachabilitydata.go [hunk 18]
```diff
@@ -0,0 +1,61 @@
...
+// is a descendent (is in the "future") of any block that previously
+// registered it.
+//
+// Note that futureCoveringTreeNodeSet is meant to be queried only if B is not
+// a reachability tree descendant of the block in question, as reachability
+// tree queries are always O(1).
+//
+// See insertNode, hasAncestorOf, and reachabilityTree.isInPast for further
+// details.
+type futureCoveringTreeNodeSet orderedTreeNodeSet
```
Comment (someone235) [inline domain/consensus/model/reachabilitydata.go:61]: the type should be public, no?

File: domain/consensus/model/readonlyutxoset.go [hunk 19]
```diff
@@ -2,13 +2,14 @@ package model
...
 	Iterator() ReadOnlyUTXOSetIterator
 	Entry(outpoint *appmessage.Outpoint) *UTXOEntry
 }
 
-// ReadOnlyUTXOSetIterator ...
+// ReadOnlyUTXOSetIterator is an iterator over all entries in a
+// ReadOnlyUTXOSet
 type ReadOnlyUTXOSetIterator interface {
 	Next() bool
 	Get() (outpoint *appmessage.Outpoint, utxoEntry *UTXOEntry)
```
Comment (someone235) [inline domain/consensus/model/readonlyutxoset.go:15]: I would expect it'll be called `Value`

File: domain/consensus/model/selectedparentiterator.go [hunk 20]
```diff
@@ -0,0 +1,10 @@
+package model
+
+import "github.com/kaspanet/kaspad/util/daghash"
+
+// SelectedParentIterator is an iterator over the selected parent
+// chain of some block
+type SelectedParentIterator interface {
+	Next() bool
+	Get() *daghash.Hash
```
Comment (someone235) [inline domain/consensus/model/selectedparentiterator.go:9]: Here too

File: domain/consensus/model/utxodiff.go [hunk 21]
```diff
@@ -0,0 +1,12 @@
...
+import "github.com/kaspanet/kaspad/app/appmessage"
+
+// UTXODiff represents a diff between two UTXO Sets.
+type UTXODiff struct {
+	ToAdd    utxoCollection
+	ToRemove utxoCollection
+}
+
+// utxoCollection represents a set of UTXOs indexed by their outpoints
+type utxoCollection map[appmessage.Outpoint]*UTXOEntry
```
Comment (someone235) [inline domain/consensus/model/utxodiff.go:12]: it should be public

File: domain/consensus/model/utxoentry.go [hunk 22]
```diff
@@ -0,0 +1,12 @@
+package model
+
+// UTXOEntry houses details about an individual transaction output in a utxo
+// set such as whether or not it was contained in a coinbase tx, the blue
+// score of the block that accepts the tx, its public key script, and how
+// much it pays.
+type UTXOEntry struct {
+	amount          uint64
```
Comment (someone235) [inline domain/consensus/model/utxoentry.go:8]: all fields should be public.

**Review by someone235 — CHANGES_REQUESTED** [review]

**Review by someone235 — APPROVED** [review]

**Review by svarogg — APPROVED** [review]

**Pull request merged**